### PR TITLE
scrub: make ABORT a validate mode

### LINKED ISSUE #9288: scrub ABORT should not rewrite valid sstables
In ABORT mode scrub *rewrites* all valid sstables and ABORTs on the first validation issue, if found.
Since it doesn't fix any error in this mode, there is no point in rewriting all valid SSTables it scrubs - it can happily be read-only, and reuse the validation logic, just abort the process on first validation error.

### PR DESCRIPTION
**Pull request by bhalevy**
ABORT doesn't fix any issues it finds, but rather it aborts scrubbing on the first error.
There is no point in it re-writing all the valid SSTables it scrubs.
Instead, it can just re-use scrub/validate mode and abort on the first
error, without rewriting the valid sstables.

Fixes #9288

Test: unit(dev)

DTest: nodetool_additional_test.py:TestNodetool.{
    scrub_sstable_with_invalid_fragment_test, scrub_ks_sstable_with_invalid_fragment_test, scrub_with_multi_nodes_expect_data_rebuild_test, scrub_with_one_node_expect_data_loss_test,
    scrub_segregate_sstable_with_invalid_fragment_test, scrub_segregate_ks_sstable_with_invalid_fragment_test,
    validate_with_one_node_expect_data_loss_test, validate_sstable_with_invalid_fragment_test, validate_ks_sstable_with_invalid_fragment_test
} (dev, w/ https://github.com/scylladb/scylla-dtest/pull/2265)

### CODE REVIEW

File: compaction/compaction_manager.cc [hunk 1]
```diff
@@ -548,7 +548,9 @@ inline bool compaction_manager::maybe_stop_on_error(future<> f, stop_iteration w
         // case we won't retry.
         retry = e.retry();
         decision_msg = !retry ? stop_msg : decision_msg;
-        cmlog.info("compaction info: {}: {}", e.what(), decision_msg);
+        auto error_msg = sstring(e.what());
+        logging::log_level level = error_msg.find("failed") == sstring::npos ? logging::log_level::info : logging::log_level::error;
+        cmlog.log(level, "compaction info: {}: {}", error_msg, decision_msg);
```
Comment (denesb) [inline compaction/compaction_manager.cc:553]: Pattern-matching the error message looks pretty fragile to me.
I'd prefer a more explicit API, maybe add a new flag.

File: compaction/compaction.cc [hunk 2]
```diff
@@ -1717,14 +1717,17 @@ static future<compaction_info> scrub_sstables_validate_mode(sstables::compaction
...
 
-    clogger.info("Scrubbing in validate mode {}", sstables_list_msg);
+    clogger.info("Scrubbing in validate mode: {}", sstables_list_msg);
 
     auto permit = cf.compaction_concurrency_semaphore().make_tracking_only_permit(schema.get(), "scrub:validate", db::no_timeout);
     auto reader = sstables->make_crawling_reader(schema, permit, descriptor.io_priority, nullptr);
 
     const auto valid = co_await scrub_validate_mode_validate_reader(std::move(reader), *info);
+    if (!valid) {
+        info->invalid_sstables++;
```
Comment (denesb) [inline compaction/compaction.cc:1733]:
```
info->invalid_sstables += sstables.size();
```
Should scrub proper also bump this?

**Review by denesb — CHANGES_REQUESTED** [review]
Minor comments, LGTM overall.

File: compaction/compaction.cc (hunk 2 above)
Comment (bhalevy) [reply compaction/compaction.cc:1733]: > `info->invalid_sstables += sstables.size();`

Are the sstables merged so we don't know which of them was invalid?
(that said, it seems like we're passing a single sstable at a time anyhow.)

> Should scrub proper also bump this?

Yes

File: compaction/compaction_manager.cc (hunk 1 above)
Comment (bhalevy) [reply compaction/compaction_manager.cc:553]: yeah, it's should easy to complement the `_retry` flag.
Maybe turn it into an enum class.
(Reminds me the good old [Abort, Retry, Fail?](https://en.wikipedia.org/wiki/Abort,_Retry,_Fail%3F) days :-) )

File: compaction/compaction.cc (hunk 2 above)
Comment (denesb) [reply compaction/compaction.cc:1733]: > > `info->invalid_sstables += sstables.size();`
> 
> Are the sstables merged so we don't know which of them was invalid?

Yes there is no way to tell

> (that said, it seems like we're passing a single sstable at a time anyhow.)

Correct, but we cannot assume this as we get a `vector<sstable>` here.

File: compaction/compaction_manager.cc (hunk 1 above)
Comment (avikivity) [reply compaction/compaction_manager.cc:553]: If we have an exception, shouldn't we use error unconditionally?

If it's an expected exception, catch it (by type) and swallow it.

Comment (bhalevy) [reply compaction/compaction_manager.cc:553]: @avikivity we (ab)use the stop_compaction_exception for several purposes.  We can split it into 3 different exceptions that can be caught and handled by type.

@raphaelsc what do you think?

**Review by nyh — APPROVED** [review]
Looks good to me, but Benny, this is an antique series please rebase.

I wonder if it makes sense to not-rewrite-sstables in all modes - not just ABORT mode - and only start copying and rewriting the sstable if we detect a problem. But whether or not this is a good idea obviously depends on how many errors we expect to find - if most sstables will have errors, or just one in 100 sstables will have errors.

**Comment by bhalevy** [comment]
> Looks good to me, but Benny, this is an antique series please rebase.

Sure, no problem.

> 
> I wonder if it makes sense to not-rewrite-sstables in all modes - not just ABORT mode - and only start copying and rewriting the sstable if we detect a problem. But whether or not this is a good idea obviously depends on how many errors we expect to find - if most sstables will have errors, or just one in 100 sstables will have errors.

This is the intention of scrub quarantine mode.
The idea is to first run scrub in VALIDATE mode, that will detect any corrupt sstables and will move them to the quarantine sub-directory.
Then run scrub in SEGREGATE mode with quarantine_mode=ONLY to segregate only those sstables previously found in VALIDATE mode and been put into the quarantine.

**Comment by bhalevy** [comment]
retracting

**Pull request closed without merging**